File: src/main/java/cn/zzwtsy/pu/dao/UserDao.java
package cn.zzwtsy.pu.dao;

import cn.zzwtsy.pu.PuCampus;
import cn.zzwtsy.pu.bean.UserBean;
import cn.zzwtsy.pu.database.DataBaseHelper;
import cn.zzwtsy.pu.utils.Encryption;

import java.sql.SQLException;
import java.util.Map;

/**
 * UserDao
 *
 * @author zzwtsy
 * @since 2022/12/23
 */
public class UserDao {
    private final String tableName = "user";
    private final int md5Length = 32;

    /**
     * 获取用户信息
     *
     * @param qqId 用户qq号
     * @return {@link UserBean}
     */
    public UserBean getUserByQqId(long qqId) {
        String encryption = encryptionQqId(qqId);
        if (encryption == null || encryption.length() != md5Length) {
            return null;
        }
        UserBean userBean = new UserBean();
        String sql = "SELECT * FROM " + tableName + " WHERE qqId = ?";
        try {
            Map<Object, Object> queryMap = DataBaseHelper.executeQueryMap(sql, encryption);
            if (queryMap != null) {
                userBean.setOauthToken((String) queryMap.get("oauthToken"));
                userBean.setOauthTokenSecret((String) queryMap.get("oauthTokenSecret"));
                userBean.setQqId((String) queryMap.get("qqId"));
                userBean.setUid((String) queryMap.get("uid"));
            } else {
                return null;
            }
        } catch (SQLException e) {
            PuCampus.INSTANCE.getLogger().error("获取用户信息失败", e);
            return null;
        }
        return userBean;
    }

    /**
     * 添加用户信息
     *
     * @param oauthToken       oauthToken
     * @param oauthTokenSecret oauthTokenSecret
     * @return 受影响的行数: >=1 添加成功
     */
    public int addUser(long qqId, String uid, String oauthToken, String oauthTokenSecret) {
        String encryption = encryptionQqId(qqId);
        if (encryption == null || encryption.length() != md5Length) {
            return -1;
        }
        int status;
        String sql = "INSERT INTO " + tableName + " (qqId,uid,oauthToken,oauthTokenSecret) VALUES (?,?,?,?)";
        try {
            status = DataBaseHelper.executeUpdate(sql, encryption, uid, oauthToken, oauthTokenSecret);
        } catch (SQLException e) {
            PuCampus.INSTANCE.getLogger().error("添加用户信息失败", e);
            return -1;
        }
        return status;
    }

    /**
     * 删除用户
     *
     * @param qqId 用户qq号
     * @return 受影响的行数: >=1 删除成功
     */
    public int deleteUser(long qqId) {
        String encryption = encryptionQqId(qqId);
        if (encryption == null || encryption.length() != md5Length) {
            return -1;
        }
        int deleteStatus;
        String sql = "DELETE FROM" + tableName + "WHERE qqId = ?";
        try {
            deleteStatus = DataBaseHelper.executeUpdate(sql, encryption);
        } catch (SQLException e) {
            PuCampus.INSTANCE.getLogger().error("删除用户信息失败", e);
            return -1;
        }
        return deleteStatus;
    }

    /**
     * 更新用户信息
     *
     * @param qqId             qq号
     * @param oauthToken       oauthToken
     * @param oauthTokenSecret oauthTokenSecret
     * @return 受影响的行数: >=1 更新成功
     */
    public int updateUser(long qqId, String uid, String oauthToken, String oauthTokenSecret) {
        String encryption = encryptionQqId(qqId);
        if (encryption == null || encryption.length() != md5Length) {
            return -1;
        }
        int status;
        String sql = "UPDATE " + tableName + " SET uid = ?,oauthToken = ?,oauthTokenSecret = ? WHERE qqId = ?";
        try {
            status = DataBaseHelper.executeUpdate(sql, uid, oauthToken, oauthTokenSecret, encryption);
        } catch (SQLException e) {
            PuCampus.INSTANCE.getLogger().error("更新用户信息失败", e);
            return -1;
        }
        return status;
    }

    /**
     * 加密 qq 号
     *
     * @param qqId qq 号
     * @return {@link String}
     */
    private String encryptionQqId(long qqId) {
        if (qqId == 0) {
            String id = "0";
            return Encryption.encryptionToMd5(id, id);
        }
        String id = new StringBuilder(String.valueOf(qqId)).reverse().toString();
        String salt = id.substring(0, 4);
        return Encryption.encryptionToMd5(id, salt);
    }
}

File: src/main/kotlin/cn/zzwtsy/pu/dao/UserDao.kt
package cn.zzwtsy.pu.dao

import cn.zzwtsy.pu.model.User
import cn.zzwtsy.pu.model.users
import cn.zzwtsy.pu.tools.ToolsKt
import org.ktorm.database.Database
import org.ktorm.dsl.eq
import org.ktorm.entity.add
import org.ktorm.entity.find
import org.ktorm.entity.update

/**
 * 用户 Dao
 * @author zzwtsy
 * @date 2023/03/21
 */
object UserDao {
    private lateinit var database: Database

    /**
     * 根据用户 qq 号获取用户信息
     * @param [qqId] qq id
     * @return [User?]
     */
    fun getUserByQqId(qqId: Long): User? {
        val md5 = ToolsKt.encryptionQqIdToMd5(qqId)
        return database.users.find { it.qqId eq md5 }
    }

    /**
     * 添加用户
     * @param [qqId] qq id
     * @param [uid] uid
     * @param [oauthToken] oauthToken
     * @param [oauthTokenSecret] oauthTokenSecret
     * @return [Int]
     */
    fun addUser(qqId: Long, uid: String, oauthToken: String, oauthTokenSecret: String): Int {
        val md5 = ToolsKt.encryptionQqIdToMd5(qqId)
        val user = User(md5, uid, oauthToken, oauthTokenSecret)
        return database.users.add(user)
    }

    /**
     * 删除用户
     * @param [qqId] qq id
     * @return [Int]
     */
    fun deleteUser(qqId: Long): Int {
        val md5 = ToolsKt.encryptionQqIdToMd5(qqId)
        return database.users.find { it.qqId eq md5 }?.delete()!!
    }

    /**
     * 更新用户信息
     * @param [qqId] qq id
     * @param [uid] uid
     * @param [oauthToken] oauthToken
     * @param [oauthTokenSecret] oauthTokenSecret
     * @return [Int]
     */
    fun updateUser(qqId: Long, uid: String, oauthToken: String, oauthTokenSecret: String): Int {
        val md5 = ToolsKt.encryptionQqIdToMd5(qqId)
        val user = User(md5, uid, oauthToken, oauthTokenSecret)
        return database.users.update(user)
    }
}